lib/reek/spec/should_reek_of.rb

```diff
@@ -17,22 +23,42 @@ def initialize(smell_category,
 
       def matches?(actual)
         self.examiner = Examiner.new(actual, configuration: configuration)
-        self.all_smells = examiner.smells
-        all_smells.any? { |warning| warning.matches?(smell_category, smell_details) }
+
+        set_failure_messages_for_smell_type && return unless matching_smell_types?
+        set_failure_messages_for_smell_details && return unless matching_smell_details?
+
+        true
```

Comment:
But the code you suggest is even more confusing, at least for me:

```Ruby
def set_failure_messages
  set_failure_messages_for_smell_details unless matching_smell_details?
  set_failure_messages_for_smell_type unless matching_smell_types?
end
```

This code gives the impression that both validations are independent from each other. But they are not. The second one is dependent on the first one because it can't stand on it's own.

With my implementation at least this is clear because I say "&& return". With your suggestion, it is (quoting a famous man) "baroque" ;)